File: internal/paste/crypto_xlang_test.go
package paste

import (
	"encoding/base64"
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

type xlangFixture struct {
	Name          string          `json:"name"`
	KeyB64Url     string          `json:"key_b64url"`
	Plaintext     json.RawMessage `json:"plaintext"`
	CiphertextB64 string          `json:"ciphertext_b64"`
	IvB64         string          `json:"iv_b64"`
}

func TestCryptoXLangFixtures(t *testing.T) {
	data, err := os.ReadFile(filepath.Join("testdata", "xlang_fixtures.json"))
	if err != nil {
		t.Fatal(err)
	}
	var fixtures []xlangFixture
	if err := json.Unmarshal(data, &fixtures); err != nil {
		t.Fatal(err)
	}
	if len(fixtures) == 0 {
		t.Fatal("no fixtures loaded")
	}
	for _, f := range fixtures {
		t.Run(f.Name, func(t *testing.T) {
			key, err := base64UrlDecode(f.KeyB64Url)
			if err != nil {
				t.Fatal(err)
			}
			ct, _ := base64.StdEncoding.DecodeString(f.CiphertextB64)
			iv, _ := base64.StdEncoding.DecodeString(f.IvB64)
			var got json.RawMessage
			if err := DecryptJSON(ct, iv, key, &got); err != nil {
				t.Fatalf("decrypt: %v", err)
			}
			var a, b any
			_ = json.Unmarshal(f.Plaintext, &a)
			_ = json.Unmarshal(got, &b)
			if !reflect.DeepEqual(a, b) {
				t.Errorf("plaintext mismatch:\nwant %s\ngot  %s", f.Plaintext, got)
			}
		})
	}
}

func TestCryptoXLangRoundtrip(t *testing.T) {
	// For each fixture, also verify that re-encrypting and re-decrypting in Go
	// produces the same plaintext (catches Go-internal regressions).
	data, _ := os.ReadFile(filepath.Join("testdata", "xlang_fixtures.json"))
	var fixtures []xlangFixture
	_ = json.Unmarshal(data, &fixtures)
	for _, f := range fixtures {
		t.Run(f.Name+"-roundtrip", func(t *testing.T) {
			key, _ := base64UrlDecode(f.KeyB64Url)
			ct, iv, err := EncryptJSON(json.RawMessage(f.Plaintext), key)
			if err != nil {
				t.Fatal(err)
			}
			var out json.RawMessage
			if err := DecryptJSON(ct, iv, key, &out); err != nil {
				t.Fatal(err)
			}
		})
	}
}

func base64UrlDecode(s string) ([]byte, error) {
	switch len(s) % 4 {
	case 2:
		s += "=="
	case 3:
		s += "="
	}
	return base64.URLEncoding.DecodeString(s)
}

File: web/src/lib/paste/crypto.xlang.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { importKey, decryptJSON } from './crypto';

const FIXTURES_PATH = join(__dirname, '../../../../internal/paste/testdata/xlang_fixtures.json');
const fixtures: Array<{
	name: string;
	key_b64url: string;
	plaintext: unknown;
	ciphertext_b64: string;
	iv_b64: string;
}> = JSON.parse(readFileSync(FIXTURES_PATH, 'utf8'));

describe('crypto xlang', () => {
	for (const f of fixtures) {
		it(`decrypts Go-produced fixture: ${f.name}`, async () => {
			const key = await importKey(f.key_b64url);
			const blob = stdB64Decode(f.ciphertext_b64);
			const iv = stdB64Decode(f.iv_b64);
			const got = await decryptJSON(blob, iv, key);
			expect(got).toEqual(f.plaintext);
		});
	}
});

function stdB64Decode(s: string): Uint8Array {
	const bin = atob(s);
	const out = new Uint8Array(bin.length);
	for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
	return out;
}
